### PR TITLE
User profile

### PR DESCRIPTION
**Pull request by reid-andrew**
Code for User Profile Page Stories 19-22
- Allow user to access Profile Page
- Allow user to edit profile data
- Allow user to edit Password
- Require email address to be unique on edit

### CODE REVIEW

File: app/controllers/passwords_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class PasswordsController < ApplicationController
```
Comment (reid-andrew) [inline app/controllers/passwords_controller.rb:1]: Made the decision to move password edit functionality into it's own controller to maintain RESTfulness. Can revisit this if we think should go in a different direction.

File: app/controllers/profile_controller.rb [hunk 2]
```diff
@@ -6,5 +6,32 @@ def require_user
...
+    if User.where(email:@user.email).where.not(id:@user.id) != []
+      flash[:failure] = "That email already exists"
+      render :edit
+    elsif @user.save
+      redirect_to '/profile'
+      flash[:error] = "Your profile is updated"
+    else
+      flash[:failure] = "You are missing required fields."
+      render :edit
+    end
```
Comment (reid-andrew) [inline app/controllers/profile_controller.rb:29]: User Story 20. Didn't call for validating all data was present, but we added in code similar to what was done at user creation to maintain consistency.

File: app/controllers/profile_controller.rb [hunk 3]
```diff
@@ -6,5 +6,32 @@ def require_user
...
+  end
+
+  def edit
+    @user = current_user
+  end
+
+  def update
+    @user = current_user
+    @user.update(user_params)
+    if User.where(email:@user.email).where.not(id:@user.id) != []
```
Comment (mlynch5187) [inline app/controllers/profile_controller.rb:20]: Def some opportunities to refactor here, we can take a look at the user-login branch for examples

File: config/routes.rb [hunk 4]
```diff
@@ -39,6 +39,11 @@
   get "/login", to: "sessions#new"
 
   get "/profile", to: "profile#index"
+  get "/profile/:id/password", to: "passwords#edit"
+  get "/profile/:id/edit", to: "profile#edit"
+  patch "/profile/:id/edit", to: "profile#update"
+  patch "/profile/:id/password", to: "passwords#update"
+
```
Comment (reid-andrew) [inline config/routes.rb:49]: Note: We didn't namespace the 'passwords' routes. May want to consider doing so for consistency with other profile/user access.

**Review by mlynch5187 — APPROVED** [review]
Looks good, potential for refactoring the update method in profile controller.

File: app/controllers/passwords_controller.rb (hunk 1 above)
Comment (mlynch5187) [reply app/controllers/passwords_controller.rb:1]: OK I wasn't sure if the password controller was necessary, but if it works, I agree we can always revisit.

**Review by javier-aguilar — APPROVED** [review]
Everything looks solid from what I'm seeing. It also looks like we implemented the same changes in regards to the password/bycrypt stuff.

**Pull request merged**